dotnet/src/Connectors/Connectors.AI.OpenAI/ImageGeneration/AzureOpenAIImageGeneration.cs

```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Linq;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.AI.ImageGeneration;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI.CustomClient;
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Text;
+
+namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI.ImageGeneration;
+
+/// <summary>
+/// Azure OpenAI Image generation
+/// <see herf="https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference#image-generation" />
+/// </summary>
+public class AzureOpenAIImageGeneration : OpenAIClientBase, IImageGeneration
+{
+    /// <summary>
+    /// Azure OpenAI Endpoint ApiVersion
+    /// </summary>
+    private readonly string _apiVersion;
+
+    /// <summary>
+    /// Generation Image Operation path
+    /// </summary>
+    private const string GenerationImageOperation = "openai/images/generations:submit";
+
+    /// <summary>
+    /// Get Image Operation path
+    /// </summary>
+    private const string GetImageOperation = "openai/operations/images";
+
+    /// <summary>
+    /// Maximum number of attempts to retrieve the image generation operation result.
+    /// </summary>
+    private readonly int _maxRetryCount;
+
+    /// <summary>
+    /// Azure OpenAI REST API endpoint
+    /// </summary>
+    private readonly string _endpoint;
+
+    /// <summary>
+    /// Azure OpenAI API key
+    /// </summary>
+    private readonly string _apiKey;
+
+    /// <summary>
+    /// Create a new instance of Azure OpenAI image generation service
+    /// </summary>
+    /// <param name="endpoint">Azure OpenAI deployment URL, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart</param>
+    /// <param name="apiKey">Azure OpenAI API key, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart</param>
+    /// <param name="httpClient">Custom <see cref="HttpClient"/> for HTTP requests.</param>
+    /// <param name="logger">Application logger</param>
+    /// <param name="maxRetryCount"> Maximum number of attempts to retrieve the image generation operation result.</param>
+    /// <param name="apiVersion">Azure OpenAI Endpoint ApiVersion</param>
+    public AzureOpenAIImageGeneration(string endpoint, string apiKey, HttpClient? httpClient = null, ILogger? logger = null, int maxRetryCount = 5, string apiVersion = "2023-06-01-preview") : base(httpClient, logger)
+    {
+        this._endpoint = endpoint;
+        this._apiKey = apiKey;
+        this._maxRetryCount = maxRetryCount;
+        this._apiVersion = apiVersion;
+    }
```

Comment:
There are a few considerations regarding the endpoint and custom HTTP client:
1. It should be possible to use endpoint provided through BaseAddress property of the custom HTTP client. The scenario is possible when a hosting application has/creates a dedicated and preconfigured HTTP client per host.
2. At the same time, if the BaseAddress property is not set then the endpoint parameter value should be used instead. That can be convenient if a hosting application needs to target different hosts dinamicaly.
3. If both BaseAddress and endpoint properties are specified the parameter value should override BaseAddress property value. This is useful in cases where a hosting application generally targets the same host in the majority of scenarios, but there may be a rare instance where it needs to communicate with a different host.
4. If neither the BaseAddress nor the endpoint parameter is specified, an exception should be thrown to prevent the construction of the class and avoid an invalid class state.

This approach is implemented in the HugginFace connector - [the class constructor](https://github.com/microsoft/semantic-kernel/blob/4c4670acb2ffb9d16ca406a61e55cd10c96b345d/dotnet/src/Connectors/Connectors.AI.HuggingFace/TextCompletion/HuggingFaceTextCompletion.cs#LL107C5-L107C5), the [Url selection logic](https://github.com/microsoft/semantic-kernel/blob/4c4670acb2ffb9d16ca406a61e55cd10c96b345d/dotnet/src/Connectors/Connectors.AI.HuggingFace/TextCompletion/HuggingFaceTextCompletion.cs#L203)

Reply:
Is this okay?
```csharp
public AzureOpenAIImageGeneration(string? endpoint, string apiKey, HttpClient? httpClient = null, ILogger? logger = null, int maxRetryCount = 5, string apiVersion = "2023-06-01-preview") : base(httpClient, logger)
{
    endpoint = !string.IsNullOrEmpty(endpoint) ? endpoint : httpClient?.BaseAddress?.AbsoluteUri;
    Verify.NotNullOrWhiteSpace(endpoint);
```

Reply:
The way of two constructors looks better. Let me work on this.